Component version resolusion on Minecraft version change

Ryex opened the pull request:
Allow LWJGL, Mappings, loader version, etc to update appropriately when the minecraft version is changed

~~Depends on #2528~~

Closes #562
Closes #703
Closes #1358 
Closes #2540

Ryex committed:
feat(components) recomend the correct lwjgl version for the minecraft version

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(instance components): resolve instance componants when changing minecraft version

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix misepellings

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
add better profile logging properly resolve important dependencies

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix qt5 lwjgl recommend + spelling

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Trial97 committed:
clang-format

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented on launcher/minecraft/PackProfile.h:
```diff
@@ -48,6 +48,7 @@
 
 #include "Component.h"
 #include "LaunchProfile.h"
+#include "minecraft/PackProfile_p.h"
```
Do we need this included?
I do not like the private header stuff but if we use them we should not include them inside the normal headers.
Is there a good reason why we should use them in favor of private members?

Ryex replied:
it *was* needed for the `ComponentContainer` type alias but I can just specify the full type

Trial97 commented on launcher/minecraft/PackProfile.cpp:
```diff
@@ -954,8 +988,58 @@ bool PackProfile::setComponentVersion(const QString& uid, const QString& version
+                                } else {
+                                    qCDebug(instanceProfileC) << d->m_instance->name() << "|"
+                                                              << "no compatible version for" << comp->getID() << "removing";
+                                    remove(comp->getID());
```
I would argue that in this case, we should either display a message to the user to signal an error or at least not remove it(unless maybe is disabled) 
As the user may be offline removing the modloader would cause trouble.

Trial97 commented:
seg fault when changing the version of Minecraft for a fabric instance
 ```
QRasterPaintEngine::brushOriginChanged (this=0x555557cdb0b0) at /usr/src/debug/qt6-base/qtbase/src/gui/painting/qpaintengine_raster.cpp:678
Downloading source file /usr/src/debug/qt6-base/qtbase/src/gui/painting/qpaintengine_raster.cpp
678	   s->fillFlags |= DirtyBrushOrigin;                                                                                                                                                                                                                                                                                   
(gdb) bt
#0  QRasterPaintEngine::brushOriginChanged (this=0x555557cdb0b0) at /usr/src/debug/qt6-base/qtbase/src/gui/painting/qpaintengine_raster.cpp:678
#1  0x00007ffff7c5f7c2 in QPainter::setBrushOrigin (this=0x7fffffffa338, p=<synthetic pointer>...) at /usr/src/debug/qt6-base/qtbase/src/gui/painting/qpainter.h:705
#2  QTreeView::drawBranches (this=0x55555743f950, painter=0x7fffffffa338, rect=..., index=...) at /usr/src/debug/qt6-base/qtbase/src/widgets/itemviews/qtreeview.cpp:1931
#3  0x00007ffff7c60315 in QTreeView::drawRow (this=<optimized out>, painter=0x7fffffffa338, option=..., index=...) at /usr/src/debug/qt6-base/qtbase/src/widgets/itemviews/qtreeview.cpp:1801
#4  0x00007ffff7c5eaf1 in QTreeView::drawTree (this=this@entry=0x55555743f950, painter=painter@entry=0x7fffffffa338, region=...) at /usr/src/debug/qt6-base/qtbase/src/widgets/itemviews/qtreeview.cpp:1536
#5  0x00007ffff7c5ed29 in QTreeView::paintEvent (this=0x55555743f950, event=0x7fffffffa630) at /usr/src/debug/qt6-base/qtbase/src/widgets/itemviews/qtreeview.cpp:1343
#6  0x00007ffff794bee5 in QWidget::event (this=0x55555743f950, event=0x7fffffffa630) at /usr/src/debug/qt6-base/qtbase/src/widgets/kernel/qwidget.cpp:9186
#7  0x00007ffff79e2838 in QFrame::event (this=0x55555743f950, e=0x7fffffffa630) at /usr/src/debug/qt6-base/qtbase/src/widgets/widgets/qframe.cpp:521
#8  0x00007ffff6544178 in QCoreApplicationPrivate::sendThroughObjectEventFilters (receiver=receiver@entry=0x555557330820, event=event@entry=0x7fffffffa630) at /usr/src/debug/qt6-base/qtbase/src/corelib/kernel/qcoreapplication.cpp:1281
#9  0x00007ffff78fc50c in QApplicationPrivate::notify_helper (this=<optimized out>, receiver=0x555557330820, e=0x7fffffffa630) at /usr/src/debug/qt6-base/qtbase/src/widgets/kernel/qapplication.cpp:3281
#10 0x00007ffff6544d28 in QCoreApplication::notifyInternal2 (receiver=0x555557330820, event=0x7fffffffa630) at /usr/src/debug/qt6-base/qtbase/src/corelib/kernel/qcoreapplication.cpp:1142
#11 0x00007ffff6544d7d in QCoreApplication::sendSpontaneousEvent (receiver=<optimized out>, event=<optimized out>) at /usr/src/debug/qt6-base/qtbase/src/corelib/kernel/qcoreapplication.cpp:1597
#12 0x00007ffff793c29e in QWidgetPrivate::sendPaintEvent (this=this@entry=0x555557075410, toBePainted=...) at /usr/src/debug/qt6-base/qtbase/src/widgets/kernel/qwidget.cpp:5650
#13 0x00007ffff793e28b in QWidgetPrivate::drawWidget (this=0x555557075410, pdev=0x555557350d30, rgn=..., offset=..., flags=..., sharedPainter=<optimized out>, repaintManager=<optimized out>) at /usr/src/debug/qt6-base/qtbase/src/widgets/kernel/qwidget.cpp:5600
#14 0x00007ffff795b990 in QWidgetRepaintManager::paintAndFlush (this=<optimized out>) at /usr/src/debug/qt6-base/qtbase/src/widgets/kernel/qwidgetrepaintmanager.cpp:900
#15 0x00007ffff794b998 in QWidget::event (this=0x555557b9fba0, event=0x555557ba5e60) at /usr/src/debug/qt6-base/qtbase/src/widgets/kernel/qwidget.cpp:9350
#16 0x00007ffff78fc51c in QApplicationPrivate::notify_helper (this=<optimized out>, receiver=0x555557b9fba0, e=0x555557ba5e60) at /usr/src/debug/qt6-base/qtbase/src/widgets/kernel/qapplication.cpp:3287
#17 0x00007ffff6544d28 in QCoreApplication::notifyInternal2 (receiver=0x555557b9fba0, event=event@entry=0x555557ba5e60) at /usr/src/debug/qt6-base/qtbase/src/corelib/kernel/qcoreapplication.cpp:1142
#18 0x00007ffff65450eb in QCoreApplication::sendEvent (receiver=<optimized out>, event=0x555557ba5e60) at /usr/src/debug/qt6-base/qtbase/src/corelib/kernel/qcoreapplication.cpp:1583
#19 QCoreApplicationPrivate::sendPostedEvents (receiver=0x0, event_type=0, data=0x5555565e4460) at /usr/src/debug/qt6-base/qtbase/src/corelib/kernel/qcoreapplication.cpp:1940
#20 0x00007ffff67a49ec in QCoreApplication::sendPostedEvents (receiver=0x0, event_type=0) at /usr/src/debug/qt6-base/qtbase/src/corelib/kernel/qcoreapplication.cpp:1797
#21 postEventSourceDispatch (s=0x555556801670) at /usr/src/debug/qt6-base/qtbase/src/corelib/kernel/qeventdispatcher_glib.cpp:244
#22 0x00007ffff5514a89 in g_main_dispatch (context=0x7fffec000f00) at ../glib/glib/gmain.c:3344
#23 0x00007ffff55769b7 in g_main_context_dispatch_unlocked (context=0x7fffec000f00) at ../glib/glib/gmain.c:4152
#24 g_main_context_iterate_unlocked.isra.0 (context=context@entry=0x7fffec000f00, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../glib/glib/gmain.c:4217
#25 0x00007ffff5513f95 in g_main_context_iteration (context=0x7fffec000f00, may_block=1) at ../glib/glib/gmain.c:4282
#26 0x00007ffff67a2cbd in QEventDispatcherGlib::processEvents (this=0x55555684c7a0, flags=...) at /usr/src/debug/qt6-base/qtbase/src/corelib/kernel/qeventdispatcher_glib.cpp:394
#27 0x00007ffff654f01e in QEventLoop::processEvents (this=0x7fffffffb010, flags=...) at /usr/src/debug/qt6-base/qtbase/src/corelib/kernel/qeventloop.cpp:100
#28 QEventLoop::exec (this=0x7fffffffb010, flags=...) at /usr/src/debug/qt6-base/qtbase/src/corelib/kernel/qeventloop.cpp:182
#29 0x0000555555afa953 in Meta::Index::getLoadedVersion (this=0x555556a08130, uid=..., version=...) at /home/trial/Projects/PrismLauncher/launcher/meta/Index.cpp:159
#30 0x000055555599bbbc in Component::getVersionFile (this=0x555557b5fe60) at /home/trial/Projects/PrismLauncher/launcher/minecraft/Component.cpp:95
#31 0x000055555599c46d in Component::getProblemSeverity (this=0x555557b5fe60) at /home/trial/Projects/PrismLauncher/launcher/minecraft/Component.cpp:239
#32 0x00005555559a6b6b in PackProfile::data (this=0x55555688f3a0, index=..., role=1) at /home/trial/Projects/PrismLauncher/launcher/minecraft/PackProfile.cpp:558
#33 0x00007ffff6767253 in QAbstractProxyModel::data (this=0x5555574aca70, proxyIndex=..., role=1) at /usr/src/debug/qt6-base/qtbase/src/corelib/itemmodels/qabstractproxymodel.cpp:311
#34 0x0000555555c1fcb6 in IconProxy::data (this=0x5555574aca70, proxyIndex=..., role=1) at /home/trial/Projects/PrismLauncher/launcher/ui/pages/instance/VersionPage.cpp:86
#35 0x00007ffff6785e21 in QSortFilterProxyModel::data (this=<optimized out>, index=..., role=1) at /usr/src/debug/qt6-base/qtbase/src/corelib/itemmodels/qsortfilterproxymodel.cpp:2209
#36 0x00007ffff674d011 in QAbstractItemModel::multiData (this=0x55555704c620, index=..., roleDataSpan=...) at /usr/src/debug/qt6-base/qtbase/src/corelib/itemmodels/qabstractitemmodel.cpp:3681
#37 0x00007ffff7c0cd41 in QModelIndex::multiData (this=<optimized out>, roleDataSpan=...) at /usr/src/debug/qt6-base/qtbase/src/corelib/itemmodels/qabstractitemmodel.h:496
#38 QStyledItemDelegate::initStyleOption (this=0x55555679b610, option=0x7fffffffb5e0, index=...) at /usr/src/debug/qt6-base/qtbase/src/widgets/itemviews/qstyleditemdelegate.cpp:263
#39 0x00007ffff7c0f092 in QStyledItemDelegate::sizeHint (this=0x55555679b610, option=..., index=...) at /usr/src/debug/qt6-base/qtbase/src/widgets/itemviews/qstyleditemdelegate.cpp:399
#40 0x00007ffff7c67704 in QTreeView::indexRowSizeHint (this=<optimized out>, index=...) at /usr/src/debug/qt6-base/qtbase/src/widgets/itemviews/qtreeview.cpp:3037
#41 0x00007ffff7c7253c in QTreeViewPrivate::itemHeight (this=0x5555570f3640, item=<optimized out>) at /usr/src/debug/qt6-base/qtbase/src/widgets/itemviews/qtreeview.cpp:3555
#42 0x00007ffff7c68c53 in QTreeViewPrivate::itemHeight (this=0x5555570f3640, item=<optimized out>) at /usr/src/debug/qt6-base/qtbase/src/widgets/itemviews/qtreeview.cpp:3548
```

Trial97 commented on launcher/minecraft/PackProfile.cpp:
```diff
@@ -954,8 +988,58 @@ bool PackProfile::setComponentVersion(const QString& uid, const QString& version
+                    } else {
+                        qCDebug(instanceProfileC) << d->m_instance->name() << "|" << comp->getID() << ":" << comp->getVersion()
+                                                  << "linked to important component: " << uid << " | Removing so it re-resolves";
+                        remove(comp->getID());
```
The removal of the component usually creates a visual glitch if you are on the Versions page where the component removed is still there but without a version.

Ryex replied:
I'm not sure what the alternative is supposed to be, personally I haven't seen this visual glitch because it get re-resolved almost instantly

Trial97 replied:
Yeah I'm not experiencing this as often as before you can ignore it

Trial97 reviewed (commented):
Other than the sig fault mentioned here are some comments:

Trial97 commented on launcher/minecraft/Component.cpp:
```diff
@@ -103,7 +91,8 @@ std::shared_ptr<class VersionFile> Component::getVersionFile() const
         if (!m_metaVersion->isLoaded()) {
-            m_metaVersion->load(Net::Mode::Online);
+            // this method is const but the loading of meta changes the information
+            APPLICATION->metadataIndex()->getLoadedVersion(m_metaVersion->uid(), m_metaVersion->version());
```
Removing this fixed the segfault

Ryex replied:
This if from the Parent PR.
this kinda needs to stay, or rather it' behavior when called form `Component::updateCachedData` needs to stay.

is the issue because the function is mark `const`? could we un-const the function? likely no as it's used in some other const functions.

could you look into fixing this in your PR?


EDIT:

![image](https://github.com/PrismLauncher/PrismLauncher/assets/508861/a20a520b-4f1f-4f31-a058-49b5a10a86df)

event loop recursion?

Trial97 replied:
If I can reproduce it in my PR, let me fix it there.
Also, I don't think is const-related see the sevfault mentioned here: https://github.com/PrismLauncher/PrismLauncher/pull/2547#issuecomment-2188686634

Trial97 replied:
yep this is on my side so going to need to fix it in the parent PR

Trial97 replied:
I removed the lines from there and it works: the m_metaVersion is set with a nonloaded version only inside ComponentUpdateTask and there it is also resolved so removing that from there should not impact

Trial97 commented:
found a small bug: 
![image](https://github.com/PrismLauncher/PrismLauncher/assets/42720688/bc643cff-3764-4ca1-899c-13214f0bf2cf)

I had started around 1.20.1 with quilt and stressed out change version stuff until I hit a version where the mappings were not available. I did get a red mark there (expected to be fully removed or something) and I was ok let's  go back to a version that should have the mappings, but no luck the mappings did not change.
Here is the current log:
```
456.713 D | [launcher.task.net.metacache.http]: Saving metacache with 20 entries
[Thread 0x7fffb37fe6c0 (LWP 56226) exited]
   457.091 D | Change "net.minecraft" to "1.19.4"
   457.091 D | [launcher.instance.profile]: "1.20.2" | scanning "net.minecraft" : "23w13a" for tree link
   457.091 D | [launcher.instance.profile]: "1.20.2" | scanning "net.fabricmc.intermediary" : "" for tree link
   457.091 D | [launcher.instance.profile]: "1.20.2" | scanning "org.quiltmc.quilt-loader" : "0.26.0-beta.1" for tree link
   457.091 D | [launcher.instance.profile]: "1.20.2" | "net.minecraft" : "23w13a" has 1 dependencies
   457.091 D | [launcher.instance.profile]: "1.20.2" | "net.minecraft" depends on "org.lwjgl3"
   457.092 D | "1.20.2" | Component list save is scheduled
   457.092 D | [launcher.task]: Task "ComponentUpdateTask(0x7fffe40183a0 ID: b1e1774e-ac1e-4251-8357-710ac7f140b2)" starting for the first time
   457.092 D | [launcher.instance.profile.resolve]: Loading components
   457.092 D | [launcher.instance.profile.resolve]: "Minecraft" is already loaded
   457.092 D | [launcher.task]: Task "SequentialTask(Load meta for net.fabricmc.intermediary:1.13-pre10 ID: 16d517de-077d-4f1a-99fc-aa63dbc968b5)" starting for the first time
   457.092 D | [launcher.instance.profile.resolve]: "1.20.2" | Remote loading is being run for "net.fabricmc.intermediary"
   457.092 D | [launcher.instance.profile.resolve]: "Quilt Loader" is already loaded
   457.092 D | [launcher.task]: Task "ComponentUpdateTask(0x55555714a450 ID: 36ed4e5b-dd3c-435c-afe0-a43494e5c8de)" starting for the first time
   457.092 D | [launcher.instance.profile.resolve]: Loading components
   457.092 D | [launcher.instance.profile.resolve]: "Minecraft" is already loaded
   457.092 D | [launcher.task]: Task "SequentialTask(Load meta for net.fabricmc.intermediary:1.13-pre10 ID: f72d05d6-cd7d-4238-83b9-cdfcb04c952f)" starting for the first time
   457.092 D | [launcher.instance.profile.resolve]: "1.20.2" | Remote loading is being run for "net.fabricmc.intermediary"
   457.092 D | [launcher.instance.profile.resolve]: "Quilt Loader" is already loaded
   457.092 D | [launcher.instance.profile]: "1.20.2" | Applying "net.minecraft" GOOD
   457.092 D | [launcher.instance.profile]: "1.20.2" | Applying "net.fabricmc.intermediary" ERROR
   457.092 D | [launcher.instance.profile]: "1.20.2" | Applying "org.quiltmc.quilt-loader" GOOD
   457.095 D | [launcher.task]: Task "Meta::BaseEntityLoadTask(0x555557a0b7a0 ID: f95e0c35-63c2-4c8e-ba7a-9489396a0e50)" starting for the first time
   457.095 D | [launcher.task]: Task "Meta::BaseEntityLoadTask(0x555557a0b7a0 ID: f95e0c35-63c2-4c8e-ba7a-9489396a0e50)" succeeded
   457.095 D | [launcher.task]: Task "Meta::BaseEntityLoadTask(0x5555581a4fa0 ID: 74d3aac7-dced-4b4e-91cc-44043c81e3f9)" starting for the first time
   457.095 D | [launcher.task]: Task "Meta::BaseEntityLoadTask(0x5555581a4fa0 ID: 74d3aac7-dced-4b4e-91cc-44043c81e3f9)" succeeded
   457.096 D | [launcher.task]: Task "Meta::BaseEntityLoadTask(0x555556821810 ID: 6f4ef3d7-dcb4-4e15-aad7-d8852807b17f)" starting for the first time
   457.096 D | [launcher.task]: Task "NetJob(Download of meta file net.fabricmc.intermediary/1.13-pre10.json ID: 65f8d4b1-b06f-476d-91b6-feacac51048e)" starting for the first time
   457.097 D | [launcher.task]: Task "Net::Download(CACHE:https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json ID: 73f6d3c6-50c5-4c7a-b643-77c9ceca294b)" starting for the first time
   457.097 D | Setting up api download
   457.097 D | [launcher.task.net.download]: "{73f6d3c6-50c5-4c7a-b643-77c9ceca294b}" Runninng  "https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json"
   457.122 C | [launcher.task.net.download]: "{73f6d3c6-50c5-4c7a-b643-77c9ceca294b}" Failed "https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json" with reason QNetworkReply::ContentNotFoundError
   457.122 C | [launcher.task.net.download]: "{73f6d3c6-50c5-4c7a-b643-77c9ceca294b}" HTTP Status 404 ;error "Error transferring https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json - server replied: "
   457.122 D | [launcher.task.net.download]: "{73f6d3c6-50c5-4c7a-b643-77c9ceca294b}" Request failed in previous step: "https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json"
   457.123 D | [launcher.task]: Task "Net::Download(CACHE:https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json ID: 73f6d3c6-50c5-4c7a-b643-77c9ceca294b)" restarting for after failing at first
   457.123 D | Setting up api download
   457.123 D | [launcher.task.net.download]: "{73f6d3c6-50c5-4c7a-b643-77c9ceca294b}" Runninng  "https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json"
   457.148 C | [launcher.task.net.download]: "{73f6d3c6-50c5-4c7a-b643-77c9ceca294b}" Failed "https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json" with reason QNetworkReply::ContentNotFoundError
   457.148 C | [launcher.task.net.download]: "{73f6d3c6-50c5-4c7a-b643-77c9ceca294b}" HTTP Status 404 ;error "Error transferring https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json - server replied: "
   457.148 D | [launcher.task.net.download]: "{73f6d3c6-50c5-4c7a-b643-77c9ceca294b}" Request failed in previous step: "https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json"
   457.149 D | [launcher.task]: Task "Net::Download(CACHE:https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json ID: 73f6d3c6-50c5-4c7a-b643-77c9ceca294b)" restarting for after failing at first
   457.149 D | Setting up api download
   457.149 D | [launcher.task.net.download]: "{73f6d3c6-50c5-4c7a-b643-77c9ceca294b}" Runninng  "https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json"
   457.175 C | [launcher.task.net.download]: "{73f6d3c6-50c5-4c7a-b643-77c9ceca294b}" Failed "https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json" with reason QNetworkReply::ContentNotFoundError
   457.175 C | [launcher.task.net.download]: "{73f6d3c6-50c5-4c7a-b643-77c9ceca294b}" HTTP Status 404 ;error "Error transferring https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json - server replied: "
   457.175 D | [launcher.task.net.download]: "{73f6d3c6-50c5-4c7a-b643-77c9ceca294b}" Request failed in previous step: "https://meta.prismlauncher.org/v1/net.fabricmc.intermediary/1.13-pre10.json"
   457.175 C | [launcher.task]: Task "NetJob(Download of meta file net.fabricmc.intermediary/1.13-pre10.json ID: 65f8d4b1-b06f-476d-91b6-feacac51048e)" failed:  "One or more subtasks failed"
   457.175 C | [launcher.task]: Task "Meta::BaseEntityLoadTask(0x555556821810 ID: 6f4ef3d7-dcb4-4e15-aad7-d8852807b17f)" failed:  "One or more subtasks failed"
   457.176 C | [launcher.task]: Task "SequentialTask(Load meta for net.fabricmc.intermediary:1.13-pre10 ID: f72d05d6-cd7d-4238-83b9-cdfcb04c952f)" failed:  "One or more subtasks failed"
   457.176 D | [launcher.instance.profile.resolve]: Remote task 0 failed:  "One or more subtasks failed"
   457.176 C | [launcher.task]: Task "ComponentUpdateTask(0x55555714a450 ID: 36ed4e5b-dd3c-435c-afe0-a43494e5c8de)" failed:  "Component metadata update task failed while downloading from remote server:\nOne or more subtasks failed"
   457.176 D | [launcher.instance.profile]: "1.20.2" | Component list update/resolve task failed  Reason: "Component metadata update task failed while downloading from remote server:\nOne or more subtasks failed"
   457.176 W | "One or more subtasks failed"
   461.842 D | "1.20.2" | Component list save performed now
```

Ryex committed:
move resolution into update actions in task.

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix segfaults consistently set component problems

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
@Trial97  Ok, I've moved in a completely new direction on this.

  - First up, I extended the `ProblemProvider` nature of a component to allow the update task to explicitly set problems for display.
  - Second, moved this secondary tree based component resolution into the update task as a "UpdateActions" step
  - Third, Propagate problems in dependencies into the parent components with a finalize step

Ryex committed:
Warn about known conflicting modloaders

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Trial97 commented on launcher/minecraft/Component.h:
```diff
@@ -16,13 +19,49 @@ class VersionList;
+    ModPlatform::ModLoaderType type;
+    QStringList knownConfictingComponents;
+};
+static const QMap<QString, ModloaderMapEntry> KNOWN_MODLOADERS{
+    { "net.neoforged", { ModPlatform::NeoForge, { "net.minecraftforge", "net.fabricmc.fabric-loader", "net.fabricmc.fabric-loader" } } },
+    { "net.minecraftforge", { ModPlatform::Forge, { "net.neoforged", "net.fabricmc.fabric-loader", "net.fabricmc.fabric-loader" } } },
+    { "net.fabricmc.fabric-loader", { ModPlatform::Fabric, { "net.minecraftforge", "net.neoforged", "org.quiltmc.quilt-loader" } } },
+    { "org.quiltmc.quilt-loader", { ModPlatform::Quilt, { "net.minecraftforge", "net.neoforged", "net.fabricmc.fabric-loader" } } },
+    { "com.mumfrey.liteloader", { ModPlatform::LiteLoader, {} } }
+};
```
Isn't the definition of non-trivial static constants recommended to be inside a cpp file and in the header only the declaration?

Ryex replied:
to the best of my knowledge you can not separate the definition and declaration of a `static const`. trying gives me redefinition errors

Trial97 replied:
Ok for me it worked:https://github.com/PrismLauncher/PrismLauncher/pull/2069/files#diff-ff43b34189e94500e7774fa80cec64045c4c008b84e5b59b4ee1ad45670a8c39
and https://github.com/PrismLauncher/PrismLauncher/pull/2069/files#diff-27de15ecbae4a7ebc0641e0a4919b91ba982c587b32e177d8ebeab21e03f438cR508-R512
should I revert that?

Ryex replied:
... I'm silly. of course it needs to be a class member. otherwise static means static to the translation unit ie. every inclusion of the header makes a new copy...

Trial97 commented on launcher/minecraft/ComponentUpdateTask.cpp:
```diff
@@ -512,24 +531,189 @@
+template <class... Ts>
+overload(Ts...) -> overload<Ts...>;
+
+void ComponentUpdateTask::performUpdateActions()
```
lol I think we can dismiss as is not such a big function

Trial97 commented on launcher/minecraft/ComponentUpdateTask.cpp:
```diff
@@ -512,24 +531,198 @@ void ComponentUpdateTask::resolveDependencies(bool checkOnly)
+                            component->addComponentProblem(ProblemSeverity::Error,
+                                                           QObject::tr("No compatible version of %1 found for %2 %3")
+                                                               .arg(component->getName(), lrc.parentName, lrc.version));
+                        }
```
```suggestion
                        if (!recommended) {
                            recommended = versionList->getLatestForParent(lrc.parentUid, lrc.version);
                        }
                        if (recommended) {
                            component->setVersion(recommended->version());
                            component->waitLoadMeta();
                            return;
                        } else {
                            component->addComponentProblem(ProblemSeverity::Error,
                                                           QObject::tr("No compatible version of %1 found for %2 %3")
                                                               .arg(component->getName(), lrc.parentName, lrc.version));
                        }
```

Trial97 commented on launcher/minecraft/ComponentUpdateTask.cpp:
```diff
@@ -512,24 +531,198 @@ void ComponentUpdateTask::resolveDependencies(bool checkOnly)
+                        << component->getVersion() << "updating to latest recommend or compatible with" << lrc.parentUid << lrc.version;
+                    auto versionList = APPLICATION->metadataIndex()->get(component->getID());
+                    versionList->waitToLoad();
+                    if (versionList) {
```
You already access the pointer before verifying it: this may cause segmentation faults.
Either switch the lines or disregard the if as the get should always return a non-nullptr versionList.

Trial97 committed:
format the code

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 reviewed (commented):
formated the code and some of my suggestions may appear outdated.
The code looks good but I found the following:
-  If the version is selected the explanation doesn't appear(need to click on the version then back to Minecraft)
- This is inconsistent but I observed if I have an instance with neoforge and I change mc version lower than supported: neoforge checkbox goes quickly from red to green staying green until I click on it (or change focus to another window) this was observed at mc version I'm yet to load.

Ryex committed:
review suggestions + spelling

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
> * If the version is selected the explanation doesn't appear(need to click on the version then back to Minecraft)

I believe I've fix this.

> * This is inconsistent but I observed if I have an instance with neoforge and I change mc version lower than supported: neoforge checkbox goes quickly from red to green staying green until I click on it (or change focus to another window) this was observed at mc version I'm yet to load.

After my changes I get this happening in reverse with LWJGL, it will go red but as soon as I hover the mouse it fixes itself. I can't find a cause, and I don't particularly feel like trying to chase it down in this PR.

Trial97 commented:
First this need a rebase(there is a conflict and formating stuff)
also will this solve: https://github.com/PrismLauncher/PrismLauncher/issues/703

Trial97 commented:
Do you need help with formatting the code?

Ryex commented:
> Do you need help with formatting the code?

Yeah, seems there are still differences.

Trial97 committed:
format the code

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into fix/component-version-resolusion

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
fix merge issue

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

DioEgizio reviewed (approved):
works perfectly

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into fix/component-version-resolusion

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into fix/component-version-resolusion

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

theofficialgman commented:
Anything blocking this being merged? This is one of the last blockers for 9.0

Trial97 commented:
Yes, we are waiting for input form @getchoo

ZekeZDev commented:
Had a quick look through, I am not as versed in the launcher code base, doesn't seem to be any significant issues however the incompatibility ~ has a couple edge cases regarding quilt.  

![Screenshot from 2024-09-08 12-32-10](https://github.com/user-attachments/assets/7609c1a2-3831-456d-a990-c79df3257f26)
expected (works with fabric):
![Screenshot from 2024-09-08 12-31-45](https://github.com/user-attachments/assets/cdfdb24c-e45f-42f3-99f4-6a558c8f94a6)

quilt and fabric also exibit the correct behaviour its only quilt and neoforge/forge

![image](https://github.com/user-attachments/assets/fb2bad14-cab5-4cf9-9130-6c2d0caba4a1)

I am happy to approve once this is fixed. Not sure if we still wish to wait for Getchoo as they are busy?

Ryex commented on launcher/minecraft/Component.cpp:
```diff
@@ -44,10 +44,19 @@
 #include <assert.h>
 
+const QMap<QString, ModloaderMapEntry> Component::KNOWN_MODLOADERS = {
+    { "net.neoforged", { ModPlatform::NeoForge, { "net.minecraftforge", "net.fabricmc.fabric-loader", "net.fabricmc.fabric-loader" } } },
+    { "net.minecraftforge", { ModPlatform::Forge, { "net.neoforged", "net.fabricmc.fabric-loader", "net.fabricmc.fabric-loader" } } },
```
```suggestion
    { "net.neoforged", { ModPlatform::NeoForge, { "net.minecraftforge", "net.fabricmc.fabric-loader", "org.quiltmc.quilt-loader" } } },
    { "net.minecraftforge", { ModPlatform::Forge, { "net.neoforged", "net.fabricmc.fabric-loader", "org.quiltmc.quilt-loader" } } },
```
not sure how this slipped. fabric was repeated instead of having quilt as a conflict

Ryex committed:
fix: Forge and neoforge conflict with quilt, not fabric twice

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

TheKodeToad reviewed (approved):
This is third time something like this has been implemented lol

Ryex commented:
> This is third time something like this has been implemented lol

in prism?

TheKodeToad commented:
Yep, [#577](https://togithub.com/PrismLauncher/PrismLauncher/pull/577) [#690](https://togithub.com/PrismLauncher/PrismLauncher/pull/690)

Ryex commented:
ah, well I guess I just perfected prior art then :p
